File: amulet/level/abc/_dimension.py
from abc import ABC, abstractmethod
from typing import Generic, TypeVar, Callable
from weakref import WeakValueDictionary
from threading import Lock

from amulet.data_types import DimensionId
from amulet.block import BlockStack
from amulet.biome import Biome
from amulet.selection import SelectionGroup

from ._level import LevelFriend, LevelT
from ._history import HistoryManagerLayer
from ._chunk_handle import ChunkKey, ChunkHandle
from ._raw_level import RawDimension


ChunkHandleT = TypeVar("ChunkHandleT", bound=ChunkHandle)
RawDimensionT = TypeVar("RawDimensionT", bound=RawDimension)


class Dimension(LevelFriend[LevelT], ABC, Generic[LevelT, RawDimensionT, ChunkHandleT]):
    _dimension_id: DimensionId
    _chunk_handles: WeakValueDictionary[tuple[int, int], ChunkHandleT]
    _chunk_handle_lock: Lock
    _chunk_history: HistoryManagerLayer[ChunkKey]
    _chunk_data_history: HistoryManagerLayer[bytes]
    _raw: RawDimensionT

    def __init__(
        self, level_ref: Callable[[], LevelT | None], dimension_id: DimensionId
    ) -> None:
        super().__init__(level_ref)
        self._dimension_id = dimension_id
        self._chunk_handles = WeakValueDictionary()
        self._chunk_handle_lock = Lock()
        self._chunk_history = self._l._o.history_manager.new_layer()
        self._chunk_data_history = self._l._o.history_manager.new_layer()
        self._raw = self._l.raw.get_dimension(self._dimension_id)

    @property
    def dimension_id(self) -> DimensionId:
        return self._dimension_id

    def bounds(self) -> SelectionGroup:
        """The editable region of the dimension."""
        return self._raw.bounds()

    def default_block(self) -> BlockStack:
        """The default block for this dimension"""
        return self._raw.default_block()

    def default_biome(self) -> Biome:
        """The default biome for this dimension"""
        return self._raw.default_biome()

    def chunk_coords(self) -> set[tuple[int, int]]:
        """
        The coordinates of every chunk that exists in this dimension.

        This is the combination of chunks saved to the level and chunks yet to be saved.
        """
        chunks: set[tuple[int, int]] = set(self._raw.all_chunk_coords())
        for key, state in self._chunk_history.resources_exist_map().items():
            if state:
                chunks.add((key.cx, key.cz))
            else:
                chunks.discard((key.cx, key.cz))
        return chunks

    def changed_chunk_coords(self) -> set[tuple[int, int]]:
        """The coordinates of every chunk in this dimension that have been changed since the last save."""
        return {(key.cx, key.cz) for key in self._chunk_history.changed_resources()}

    @abstractmethod
    def _create_chunk_handle(self, cx: int, cz: int) -> ChunkHandleT:
        raise NotImplementedError

    def get_chunk_handle(self, cx: int, cz: int) -> ChunkHandleT:
        key = cx, cz
        with self._chunk_handle_lock:
            chunk_handle = self._chunk_handles.get(key)
            if chunk_handle is None:
                chunk_handle = self._chunk_handles[key] = self._create_chunk_handle(
                    cx, cz
                )
            return chunk_handle

File: src/amulet/level/java/dimension.cpp
#include "dimension.hpp"
#include "chunk_handle.hpp"

namespace Amulet {

JavaDimension::JavaDimension(
    std::shared_ptr<JavaRawDimension> raw_dimension,
    HistoryManager& history_manager,
    std::shared_ptr<bool> history_enabled)
    : _raw_dimension(std::move(raw_dimension))
    , _chunk_history(history_manager.new_layer<detail::ChunkKey>())
    , _chunk_data_history(history_manager.new_layer<std::string>())
    , _history_enabled(std::move(history_enabled))
{
}

JavaDimension::~JavaDimension()
{
}

const DimensionID& JavaDimension::get_dimension_id() const
{
    return _raw_dimension->get_dimension_id();
}

std::variant<SelectionBox, SelectionGroup> JavaDimension::get_bounds() const
{
    return _raw_dimension->get_bounds();
}

const BlockStack& JavaDimension::get_default_block() const
{
    return _raw_dimension->get_default_block();
}

const Biome& JavaDimension::get_default_biome() const
{
    return _raw_dimension->get_default_biome();
}

std::shared_ptr<JavaChunkHandle> JavaDimension::get_java_chunk_handle(std::int64_t cx, std::int64_t cz)
{
    auto key = std::make_pair(cx, cz);
    {
        std::shared_lock handle_lock(_chunk_handles_mutex);
        auto it = _chunk_handles.find(key);
        if (it != _chunk_handles.end()) {
            auto chunk_handle = it->second.lock();
            if (chunk_handle) {
                return chunk_handle;
            }
        }
    }
    {
        std::lock_guard handle_lock(_chunk_handles_mutex);
        auto it = _chunk_handles.find(key);
        if (it == _chunk_handles.end()) {
            auto chunk_handle = std::shared_ptr<JavaChunkHandle>(
                new JavaChunkHandle(
                    get_dimension_id(),
                    cx,
                    cz,
                    _raw_dimension,
                    _chunk_history,
                    _chunk_data_history,
                    _history_enabled));
            _chunk_handles.emplace(key, chunk_handle);
            return chunk_handle;
        } else {
            auto chunk_handle = it->second.lock();
            if (!chunk_handle) {
                auto chunk_handle = std::shared_ptr<JavaChunkHandle>(
                    new JavaChunkHandle(
                        get_dimension_id(),
                        cx,
                        cz,
                        _raw_dimension,
                        _chunk_history,
                        _chunk_data_history,
                        _history_enabled));
                it->second = chunk_handle;
            }
            return chunk_handle;
        }
    }
}

std::shared_ptr<ChunkHandle> JavaDimension::get_chunk_handle(std::int64_t cx, std::int64_t cz)
{
    return get_java_chunk_handle(cx, cz);
}

} // namespace Amulet
